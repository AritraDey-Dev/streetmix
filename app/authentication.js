const jwt = require('express-jwt')
const config = require('config')
const jwksRsa = require('jwks-rsa')

const secret = jwksRsa.expressJwtSecret({
  cache: true,
  rateLimit: true,
  jwksRequestsPerMinute: 5,
  jwksUri: 'http://streetmix.auth0.com/.well-known/jwks.json'
})

const jwtCheck = jwt({
  algorithms: ['RS256'],
  secret,
  issuer: 'https://streetmix.auth0.com/',
  audience: config.auth0.client_id,
  credentialsRequired: false
})

module.exports = jwtCheck
