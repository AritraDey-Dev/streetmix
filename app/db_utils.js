const logger = require('../lib/logger.js')()

const combineWrites = (mainMethod, altMethod) => {
  return async function (req, res) {
    const resObject = { status: 200 }
    // alt
    const alternateResponse = {}
    alternateResponse.header = () => alternateResponse
    alternateResponse.set = () => alternateResponse
    alternateResponse.end = () => alternateResponse
    alternateResponse.status = () => alternateResponse

    // main
    const mainResponse = {}
    mainResponse.header = () => mainResponse
    mainResponse.set = (foo, bar) => {
      res.set(foo, bar)
      return mainResponse
    }
    mainResponse.status = (e) => {
      resObject.status = e
      return mainResponse
    }
    mainResponse.json = (mainData) => {
      alternateResponse.mainData = mainData
      alternateResponse.json = (altData) => {
        const comparify = (data) => {
          data.createdAt = ''
          data.updatedAt = ''

          const cleaned = {}
          Object.keys(data)
            .sort()
            .forEach((item) => {
              if (
                data[item] &&
                data[item] !== null &&
                Object.entries(data[item]).length !== 0
              ) {
                cleaned[item] = data[item]
              }
            })
          return JSON.stringify(cleaned, Object.keys(cleaned).sort())
        }
        // if resulting objects are unequal, log an alarm
        if (comparify(mainData) !== comparify(altData)) {
          logger.error(
            `main/alt response for ${req.method} ${req.url} DIFFER: \n ${JSON.stringify({
              m: comparify(mainData),
              a: comparify(altData)
            })}`
          )
        }

        return res
          .status(mainData.status || 200)
          .json(mainData)
          .end()

        // return alternateResponse
      }
      alternateResponse.json.bind(this)
      alternateResponse.send = alternateResponse.json
      alternateResponse.end = () => alternateResponse.json(resObject)
      altMethod(req, alternateResponse)

      return mainResponse
    }
    mainResponse.json.bind(this)
    mainResponse.send = mainResponse.json
    mainResponse.end = () => mainResponse.json(resObject)
    mainMethod(req, mainResponse)
  }
}

module.exports = {
  combineReads: combineWrites, // todo add read or write specific switches
  combineWrites
}
