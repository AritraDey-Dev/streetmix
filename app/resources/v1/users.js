
const config = require('config')
const uuid = require('uuid')
const User = require('../../models/user.js')
const { ERRORS } = require('../../../lib/util')
const logger = require('../../../lib/logger.js')()

exports.post = function (req, res) {
  let loginToken = null
  const handleCreateUser = function (err, user) {
    if (err) {
      logger.error(err)
      res.status(500).send('Could not create user.')
      return
    }
    const userJson = { id: user.id, loginToken: loginToken }
    logger.info({ user: userJson }, 'New user created.')
    res.header('Location', config.restapi.baseuri + '/v1/users/' + user.id)
    res.status(201).send(userJson)
  } // END function - handleCreateUser

  const handleUpdateUser = function (err, user) {
    if (err) {
      logger.error(err)
      res.status(500).send('Could not update user.')
      return
    }

    const userJson = { id: user.id, loginToken: loginToken }
    logger.info({ user: userJson }, 'Existing user issued new login token.')

    res.header('Location', config.restapi.baseuri + '/v1/users/' + user.id)
    res.status(200).send(userJson)
  } // END function - handleUpdateUser

  const handleAuth0TwitterSignIn = function (credentials) {
    const handleFindUser = function (err, user) {
      if (err) {
        logger.error(err)
        res.status(500).send('Error finding user with Auth0 ID.')
        return
      }

      loginToken = uuid.v1()
      if (!user) {
        const u = new User({
          id: credentials.screenName,
          auth0_id: credentials.auth0_id,
          login_tokens: [ loginToken ],
          profile_image_url: credentials.profile_image_url
        })
        u.save(handleCreateUser)
      } else {
        user.id = credentials.screenName
        user.auth0_id = credentials.auth0_id
        user.profile_image_url = credentials.profile_image_url
        user.login_tokens.push(loginToken)
        user.save(handleUpdateUser)
      }
    } // END function - handleFindUser

    User.findOne({ id: credentials.screenName }, handleFindUser)
  } // END function - handleAuth0TwitterSignIn

  const handleTwitterSignIn = function (twitterCredentials) {
    // TODO: Call Twitter API with OAuth access credentials to make sure they are valid
    const handleFindUser = function (err, user) {
      if (err) {
        logger.error(err)
        res.status(500).send('Error finding user with Twitter ID.')
        return
      }
      loginToken = uuid.v1()
      if (!user) {
        const u = new User({
          id: twitterCredentials.screenName,
          twitter_id: twitterCredentials.userId,
          twitter_credentials: {
            access_token_key: twitterCredentials.oauthAccessTokenKey,
            access_token_secret: twitterCredentials.oauthAccessTokenSecret
          },
          login_tokens: [ loginToken ]
        })
        u.save(handleCreateUser)
      } else {
        user.id = twitterCredentials.screenName
        user.twitter_id = twitterCredentials.userId
        user.twitter_credentials = {
          access_token_key: twitterCredentials.oauthAccessTokenKey,
          access_token_secret: twitterCredentials.oauthAccessTokenSecret
        }
        user.login_tokens.push(loginToken)
        user.save(handleUpdateUser)
      }
    } // END function - handleFindUser
    // Try to find user with twitter ID
    User.findOne({ twitter_id: twitterCredentials.userId }, handleFindUser)
  } // END function - handleTwitterSignIn

  /**
   * Returns a randomly-generated 4-digit string of a number between 0000 and 9999
   *
   * @returns {string}
   */
  const generateRandomId = () => Math.floor(Math.random() * 10000).toString().padStart(4, '0')

  const generateId = function (nickname) {
    // TODO - Check if the Id generated is not existing
    const id = generateRandomId()
    return nickname + '-' + id
  }

  const handleAuth0SignIn = async function (credentials) {
    try {
      const user = await User.findOne({ auth0_id: credentials.auth0_id })
      loginToken = uuid.v1()
      if (!user) {
        const numOfUser = await User.count({ id: credentials.nickname })
        // Ensure there is no existing user with id same this nickname
        if (numOfUser === 0) {
          const u = new User({
            id: credentials.nickname,
            auth0_id: credentials.auth0_id,
            email: credentials.email,
            login_tokens: [ loginToken ],
            profile_image_url: credentials.profile_image_url
          })
          u.save(handleCreateUser)
        } else {
          const id = generateId(credentials.nickname)
          const u = new User({
            id: id,
            auth0_id: credentials.auth0_id,
            email: credentials.email,
            login_tokens: [ loginToken ],
            profile_image_url: credentials.profile_image_url
          })
          u.save(handleCreateUser)
        }
      } else {
        user.auth0_id = credentials.auth0_id
        user.profile_image_url = credentials.profile_image_url
        user.email = credentials.email
        user.login_tokens.push(loginToken)
        user.save(handleUpdateUser)
      }
    } catch (err) {
      logger.error(err)
      console.log(err)
      res.status(500).send('Error finding user with Auth0 ID.')
    }
  } // END function - handleAuth0SignIn

  let body
  try {
    body = req.body
  } catch (e) {
    res.status(400).send('Could not parse body as JSON.')
    return
  }

  if (body.hasOwnProperty('twitter')) {
    handleTwitterSignIn(body.twitter)
  } else if (body.hasOwnProperty('auth0_twitter')) {
    handleAuth0TwitterSignIn(body.auth0_twitter)
  } else if (body.hasOwnProperty('auth0')) {
    handleAuth0SignIn(body.auth0)
  } else {
    res.status(400).send('Unknown sign-in method used.')
  }
} // END function - exports.post

exports.get = async function (req, res) {
  // Flag error if user is not authenticated (loginToken)
  if (!req.loginToken) {
    res.status(400).send('Please provide login token.')
    return
  }

  let requestUser

  const findUserByLoginToken = async function (loginToken) {
    try {
      requestUser = await User.findOne({ login_tokens: { $in: [ loginToken ] } })
    } catch (err) {
      logger.error(err)
      throw new Error(ERRORS.CANNOT_GET_USER)
    }

    if (!requestUser) {
      throw new Error(ERRORS.UNAUTHORISED_ACCESS)
    }

    return requestUser
  }

  const findStreetmixUsers = async function () {
    let users

    try {
      users = await User.find({})
    } catch (err) {
      logger.error(err)
      throw new Error(ERRORS.CANNOT_GET_USER)
    }

    if (!users) {
      throw new Error(ERRORS.UNAUTHORISED_ACCESS)
    }

    return users
  }

  const handleFindUsers = function (users) {
    let usersArray = []

    const getUserJson = function (user) {
      const auth = req.loginToken && (user.login_tokens.indexOf(req.loginToken) !== -1 || requestUser.roles.includes('ADMIN'))

      user.asJson({ auth: auth }, function (err, userJson) {
        if (err) {
          logger.error(err)
          res.status(500).send('Could not render user JSON.')
          return
        }

        userJson.profileImageUrl = user.profile_image_url
        usersArray.push(userJson)
      })
    }

    users.forEach((user) => { getUserJson(user) })
    res.status(200).send(usersArray)
  }

  const handleError = function (error) {
    switch (error) {
      case ERRORS.USER_NOT_FOUND:
        res.status(404).send('User not found.')
        return
      case ERRORS.CANNOT_GET_USER:
        res.status(500).send('Error finding user.')
        return
      case ERRORS.UNAUTHORISED_ACCESS:
        res.status(401).send('User with that login token not found.')
        return
      default:
        res.status(500).end()
    }
  }

  findUserByLoginToken(req.loginToken)
    .then(findStreetmixUsers)
    .then(handleFindUsers)
    .catch(handleError)

} // END function - exports.get