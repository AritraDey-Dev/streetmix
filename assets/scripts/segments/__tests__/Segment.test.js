/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import {render, fireEvent, waitForElement, cleanup} from 'react-testing-library'

import { renderWithRedux } from '../../../../test/helpers/render'

import ConnectedSegment, { Segment } from '../Segment'
import { getSpriteDef, getSegmentInfo, getSegmentVariantInfo } from '../info'
//import { getVariantInfoDimensions } from '../view'
import { infoBubble } from '../../info_bubble/info_bubble'
import { SETTINGS_UNITS_IMPERIAL } from '../../users/constants'
import SEGMENT_INFO from '../info.json'
import {
  SEGMENT_WARNING_OUTSIDE,
  SEGMENT_WARNING_WIDTH_TOO_SMALL,
  SEGMENT_WARNING_WIDTH_TOO_LARGE
} from '../constants'
import { images } from '../../app/load_resources'

jest.mock('../../app/load_resources')
jest.mock('../info')
jest.mock('../view', () => {
  const actual = jest.requireActual('../view')
  return {
    ...actual,
    getVariantInfoDimensions: jest.fn(() => ({ left: 0, right: 100, center: 50})),
  }
})
jest.mock('../../streets/data_model', () => {})

jest.mock('react-transition-group', () => {
  return {
    CSSTransition: jest.fn(({children, in: show}) => (show ? children : null))
  };
});
jest.mock('react-dnd', () => {
  return {
    DragSource: (d) => (jest.fn((e) => e)),
    DropTarget: (d) => (jest.fn((e) => e)),
  };
});

const mockFetchPromise = Promise.resolve({ // 3
});

function connectDropTarget (el) { return el }
function connectDragSource (el) { return el }
const connectDragPreview = () => { }

describe('Segment', () => {
  const updateSegmentData = jest.fn()
  let segment = {}
  beforeEach(() => {
    const variant = {}
    const segment = { name: 'Segment', nameKey: 'key', zIndex: 1 }
    getSegmentInfo.mockImplementation(() => segment)
    getSegmentVariantInfo.mockImplementation(() => variant)
    //getVariantInfoDimensions.mockImplementation(() => ({ left: 0, right: 100, center: 50}))
  })
  describe('on mount', () => {
    it('updates the left position for this segment', () => {
      shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} />)
      expect(updateSegmentData).toHaveBeenCalledTimes(1)
    })
  })
  describe('if the segment variant is updated', () => {
    let updatePerspective
    beforeEach(() => {
      segment.variantString = 'variant'
      updatePerspective = jest.fn()
    })
    describe('variant is different', () => {
      it('updates perspective', () => {
        const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} updatePerspective={updatePerspective} />)
        wrapper.setProps({ segment: { variantString: 'other' } })
        expect(updatePerspective).toHaveBeenCalledTimes(4) // should probably only be 2
      })
    })
    it('updates the left position for this segment', () => {
      const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} updatePerspective={updatePerspective} />)
      wrapper.setProps({ segment: { variantString: 'other' } })
      expect(updateSegmentData).toHaveBeenCalledTimes(7) // should probably only be 1
    })
  })
  describe('shows the infobubble', () => {
    it('when segment removing or dragging action ends', () => {
      const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} isDragging />)
      wrapper.setProps({ isDragging: false })
      expect(infoBubble.considerShowing).toHaveBeenCalledTimes(1)
    })
    it('when the mouseEnter is suppressed', () => {
      const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} suppressMouseEnter />)
      wrapper.setProps({ suppressMouseEnter: false })
      expect(infoBubble.considerShowing).toHaveBeenCalledTimes(2) // should probably only be 1
    })
  })
  it('renders the units correctly', () => {
    const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} units={SETTINGS_UNITS_IMPERIAL} cssTransform={'transform'} segmentPos={1} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders active segment correctly', () => {
    const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} isDragging={false} dataNo={10} activeSegment={10} cssTransform={'transform'} segmentPos={1} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders segment warnings outside correctly', () => {
    segment.warnings = {}
    segment.warnings[SEGMENT_WARNING_OUTSIDE] = 'warning'
    const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} isDragging={false} dataNo={10} activeSegment={10} cssTransform={'transform'} segmentPos={1} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders segment warnings correctly', () => {
    segment.warnings = {}
    segment.warnings[SEGMENT_WARNING_WIDTH_TOO_SMALL] = 'warning'
    const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} isDragging={false} dataNo={10} activeSegment={10} cssTransform={'transform'} segmentPos={1} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders segment warnings width too large correctly', () => {
    segment.warnings = {}
    segment.warnings[SEGMENT_WARNING_WIDTH_TOO_LARGE] = 'warning'
    const wrapper = shallow(<Segment connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} segment={segment} actualWidth={1} updateSegmentData={updateSegmentData} connectDragPreview={connectDragPreview} isDragging={false} dataNo={10} activeSegment={10} cssTransform={'transform'} segmentPos={1} />)
    expect(wrapper).toMatchSnapshot()
  })
  it.only('renders correctly with deep render', async () => {
    const variantString = 'inbound|regular'
    const type = 'streetcar'
    const variant = SEGMENT_INFO[type].details[variantString]
    getSegmentVariantInfo.mockImplementation(() => variant)
    getSpriteDef.mockImplementation(() => ({ id: 'markings--straight-inbound', width: 4, offsetY: 11.12 }))
    //getSpriteDef should either be unmocked or be defined
    segment =  { type , variantString , segmentType: type, id: '1', width: 400 , randSeed: 1 }
    const wrapper = renderWithRedux(<ConnectedSegment connectDragSource={connectDragSource} connectDropTarget={connectDropTarget} segment={segment} actualWidth={400} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()}/>, { initialState: { ui: { activeSegment: 0}, street: { segments: [segment]}}})
    wrapper.debug()
  })
})
