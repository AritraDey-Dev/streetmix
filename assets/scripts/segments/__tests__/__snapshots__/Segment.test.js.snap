// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment renders active segment correctly 1`] = `
<div
  className="segment hover show-drag-handles"
  data-width={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "transform": "translateX(1px)",
      "width": "12px",
      "zIndex": 1,
    }
  }
>
  <span
    className="name"
  >
    Segment
  </span>
  <span
    className="width"
  >
    <React.Memo
      units={2}
      value={1}
    />
  </span>
  <span
    className="grid units-metric"
  />
  <Connect(SegmentDragHandles)
    width={12}
  />
  <CSSTransition
    classNames="switching-away"
    in={true}
    key="old-variant"
    onExited={[Function]}
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <CSSTransition
    classNames="switching-in"
    in={false}
    key="new-variant"
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <div
    className="hover-bk"
  />
</div>
`;

exports[`Segment renders segment warnings correctly 1`] = `
<div
  className="segment hover show-drag-handles warning"
  data-width={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "transform": "translateX(1px)",
      "width": "12px",
      "zIndex": 1,
    }
  }
>
  <span
    className="name"
  >
    Segment
  </span>
  <span
    className="width"
  >
    <React.Memo
      units={2}
      value={1}
    />
  </span>
  <span
    className="grid units-metric"
  />
  <Connect(SegmentDragHandles)
    width={12}
  />
  <CSSTransition
    classNames="switching-away"
    in={true}
    key="old-variant"
    onExited={[Function]}
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <CSSTransition
    classNames="switching-in"
    in={false}
    key="new-variant"
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <div
    className="hover-bk"
  />
</div>
`;

exports[`Segment renders segment warnings outside correctly 1`] = `
<div
  className="segment hover show-drag-handles warning outside"
  data-width={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "transform": "translateX(1px)",
      "width": "12px",
      "zIndex": 1,
    }
  }
>
  <span
    className="name"
  >
    Segment
  </span>
  <span
    className="width"
  >
    <React.Memo
      units={2}
      value={1}
    />
  </span>
  <span
    className="grid units-metric"
  />
  <Connect(SegmentDragHandles)
    width={12}
  />
  <CSSTransition
    classNames="switching-away"
    in={true}
    key="old-variant"
    onExited={[Function]}
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <CSSTransition
    classNames="switching-in"
    in={false}
    key="new-variant"
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <div
    className="hover-bk"
  />
</div>
`;

exports[`Segment renders segment warnings width too large correctly 1`] = `
<div
  className="segment hover show-drag-handles warning"
  data-width={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "transform": "translateX(1px)",
      "width": "12px",
      "zIndex": 1,
    }
  }
>
  <span
    className="name"
  >
    Segment
  </span>
  <span
    className="width"
  >
    <React.Memo
      units={2}
      value={1}
    />
  </span>
  <span
    className="grid units-metric"
  />
  <Connect(SegmentDragHandles)
    width={12}
  />
  <CSSTransition
    classNames="switching-away"
    in={true}
    key="old-variant"
    onExited={[Function]}
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <CSSTransition
    classNames="switching-in"
    in={false}
    key="new-variant"
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <div
    className="hover-bk"
  />
</div>
`;

exports[`Segment renders the units correctly 1`] = `
<div
  className="segment hover show-drag-handles"
  data-width={1}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "transform": "translateX(1px)",
      "width": "12px",
      "zIndex": 1,
    }
  }
>
  <span
    className="name"
  >
    Segment
  </span>
  <span
    className="width"
  >
    <React.Memo
      units={1}
      value={1}
    />
  </span>
  <span
    className="grid units-imperial"
  />
  <Connect(SegmentDragHandles)
    width={12}
  />
  <CSSTransition
    classNames="switching-away"
    in={true}
    key="old-variant"
    onExited={[Function]}
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <CSSTransition
    classNames="switching-in"
    in={false}
    key="new-variant"
    timeout={250}
    unmountOnExit={true}
  >
    <div
      className="segment-canvas-container"
    >
      <Connect(SegmentCanvas)
        actualWidth={1}
        variantString="variant"
      />
    </div>
  </CSSTransition>
  <div
    className="hover-bk"
  />
</div>
`;
